packages/data/src/index.js

```diff
@@ -79,6 +79,9 @@ export const select = defaultRegistry.select;
  * Given the name of a registered store, returns an object of the store's action creators.
  * Calling an action creator will cause it to be dispatched, updating the state value accordingly.
  *
+ * Note: If the action creator is a generator then a promise is returned.
+ * Otherwise, the dispatch will return `undefined`.
```

Comment:
I'm starting to wonder if we'd just be better off implementing `createBoundAction` as:

```js
return Promise.resolve( store.dispatch( action( ...args ) ) );
```

Which automatically normalizes everything as a promise, avoiding any confusion about what's returned and under what circumstances.

A possible further step to make a stronger guarantee of resolving-to-`undefined`:

```js
return Promise.resolve( store.dispatch( action( ...args ) ) ).then( () => Promise.resolve() );
```

Reply:
Re: modifying createBoundAction to always return as a promise.  Ya that would help with consistency

Re: always resolving to `undefined`.  This I'm not sure about as it would break our implementation.  Is the concern mostly about returning action objects?

Note, without the redux-routine middleware applied, I observed in a test (that I didn't include here) that if a promise is returned from the action object and resolves to a non action like object, an error is thrown.

So I'm curious, what's the reasoning behind wanting to always resolve to undefined as suggested?  Is it mostly because of convention (or its a bad pattern otherwise)?  I'm asking out of concern there's something I'm ignorant on.

Reply:
> I'm curious, what's the reasoning behind wanting to always resolve to undefined as suggested?

For now, I see it as part of being more explicit about exactly what we're expecting. Because otherwise, what do we expect it to resolve to? I'm not sure it's a question we've sought to answer yet, and have deferred largely to things like action objects and raw reducer state to be internal implementation details. At least for core code, this has the benefit of reducing the maintenance burden of backwards-compatibility (anticipating and allowing for changes in reducer state, presumably also action object shape).

Reply:
> For now, I see it as part of being more explicit about exactly what we're expecting. Because otherwise, what do we expect it to resolve to? I'm not sure it's a question we've sought to answer yet, and have deferred largely to things like action objects and raw reducer state to be internal implementation details. At least for core code, this has the benefit of reducing the maintenance burden of backwards-compatibility (anticipating and allowing for changes in reducer state, presumably also action object shape).

Gotcha 👍 .  Could we say we're being explicit enough by saying you can always expect a promise?  I think it's a reasonable expectation that action like objects will always resolve to undefined but if implementors of the data api need to resolve to some other value for their own implementation that's up to them?  So the api is still fairly explicit imo:

- non generator actions must always *return* an action object.
- actions can return a promise that could resolve to anything.
- generator actions yield actions and can return anything (which will be exposed via a Promise that resolves on generator completion).

I think it's reasonable for consuming code to understand that promises returned by dispatch actions _may_ resolve to undefined, so only handle the resolved value if certain the dispatched action resolves to one.

Maintenance wise, I think it's reasonable to simply always return a promise on dispatched actions and that's the contract established.

Reply:
I'm generally leaning to being okay with all of this, at least in the sense that by most default usage, you'd expect `undefined`, except when going out of your way to do otherwise. And by the fact that this restores a behavior existing prior to Gutenberg 5.4.

A possible issue is that "going out of your way" is not really as hard as it seems. For example, what would the promise of this existing action creator resolve to? 

https://github.com/WordPress/gutenberg/blob/7dfc26842280bf5254f191bd6c3efe54e84e67ee/packages/block-editor/src/store/actions.js#L382-L411

I'm expecting it resolves to the action object.

Also as noted in Slack DM, we discussed that it may not be entirely obvious that if the promise resolves to a value which has a `type` property, whether that's an action or not, it would be passed through to `dispatch`.

Reply:
> For example, what would the promise of this existing action creator resolve to?

Good question, I think it'd be good if tests in this branch cover that kind of scenario.   I'll update the tests to cover that.

Reply:
> Good question, I think it'd be good if tests in this branch cover that kind of scenario. I'll update the tests to cover that.

Turns out they already do.  As you suspected an action object is returned.